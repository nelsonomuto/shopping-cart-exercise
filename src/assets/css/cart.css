.cart-list {
  margin-bottom: 16px;
}

.cart-item {
  display: inline-grid;
  grid-template-columns: auto 70% 60px;
  grid-template-rows: auto;
  column-gap: 10px;
  row-gap: 1em;
  border-bottom: 1px solid #eee;
  width: 100%;
  padding: 16px 0;
}

.cart-item .cart-item-image {
  align-self: center;
  height: 100%;
  display: flex;
  align-items: center;
  justify-content: center;
}
.cart-item .cart-item-image > img {
  margin: 0;
}

.cart-item .cart-item-controls {
  display: flex;
  align-items: center;
  justify-content: center;
}

.cart-item-title {
  text-overflow: ellipsis;
  overflow: hidden;
  white-space: nowrap;
}

.cart-item-price {
  font-size: 0.9em;
  font-weight: bold;
}

.cart-item-controls .cart-item-update-qty-button {
  width: 20px;
  height: 20px;
  border-radius: 50%;
  outline: 0;
  border: 0;
  padding: 0;
  color: rgb(156, 156, 156);
  line-height: normal;
  text-align: center;
  font-size: 0.9em;
  font-weight: bold;
}
.cart-item-controls .cart-item-update-qty-button:hover {
  background: #eee;
  color: #222;
}
.cart-item-controls .cart-item-qty {
  margin: 4px;
  font-size: 0.9em;
}

.cart-portal-container {
  height: 100%;
  width: 100%;
  display: flex;
  position: fixed;
  top: 0;
  left: 0;
  transition: all 0.3s ease-in-out;
}

.cart-portal-container.cart-animate-open {
  background: rgba(0, 0, 0, 0.6);
}

.cart-portal-container .cart-sidebar-container {
  width: 30em;
  background: white;
  transition: all 0.3s cubic-bezier(0.445, 0.05, 0.55, 0.95);
  height: 100vh;
  position: absolute;
  right: 0;
  top: 0;
  transform: translateX(100%);
}

.cart-portal-container.cart-animate-open .cart-sidebar-container {
  opacity: 1;
  transform: translateX(0);
}

.cart-portal-container.cart-animate-closed .cart-sidebar-container {
  opacity: 0;
  transform: translateX(100%);
  pointer-events: auto;
}

.cart-portal-container.cart-animate-closed {
  background: rgba(0, 0, 0, 0);
  pointer-events: none;
}

.cart-sidebar-container {
  display: flex;
  position: relative;
  flex-flow: column nowrap;
  align-items: stretch;
  justify-content: flex-start;
  max-height: 100vh;
  overflow: hidden;
}

.cart-portal-close-button {
  position: absolute;
  top: 10px;
  right: 10px;
  outline: 0;
  border: 0;
}

.cart-portal-close-button::after {
  width: 35px;
  height: 35px;
  content: "\D7";
  border-radius: 50%;
  background: #e4e4e4;
  display: inline-block;
  line-height: 1em;
  font-size: 30px;
  cursor: pointer;
}
.cart-portal-close-button:hover::after {
  background: #efefef;
  color: rgb(71, 171, 230);
}

.cart-sidebar-items-container {
  overflow: scroll;
  padding: 16px;
}
.cart-sidebar-summary-container {
  width: 100%;
  flex: 1 1 auto;
  padding: 16px;
  background: #ddd;
}
.cart-checkout-button {
  font-size: 1.2em;
  font-weight: bold;
  width: 100%;
  padding: 20px;
  background: rgb(0, 190, 73);
  display: flex;
  align-items: center;
  justify-content: center;
  text-transform: uppercase;
  color: #333;
}
.cart-checkout-button:hover {
  background: rgb(74, 251, 141);
  color: rgb(87, 87, 87);
}

.cart-item-price-row {
  vertical-align: text-bottom;
}
.cart-item-remove-button {
  color: rgb(11, 144, 221);
  cursor: pointer;
  outline: 0;
  border: 0;
  margin: 0 6px;
  text-decoration: none;
}
.cart-item-remove-button:hover {
  color: rgb(71, 171, 230);
  text-decoration: underline;
}

.cart-sidebar-summary-container .cart-subtotal {
  margin: 10px 0;
  text-align: right;
}
.cart-sidebar-summary-container .cart-subtotal .cart-subtotal-label {
  font-weight: bold;
  font-size: 0.8em;
  color: #999;
  text-transform: uppercase;
  line-height: 1em;
}
.cart-sidebar-summary-container .cart-subtotal .cart-subtotal-amt {
  font-weight: bold;
  font-size: 1.2em;
  color: #222;
}

.cart-summary {
  text-align: right;
  margin: 20px auto;
}

.cart-summary .cart-summary-total {
  font-weight: bold;
}
