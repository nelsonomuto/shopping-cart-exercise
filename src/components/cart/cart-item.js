import React from "react"

const CartItem = ({ item, onRemove, onIncrement, onDecrement }) => (
  <div>
    <h4 className="cart-item-title">{item.title}</h4>
    <div className="cart-item-controls">
      <button className="cart-item-update-qty-button" onClick={onDecrement}>
        -
      </button>
      <span className="cart-item-qty">{item.qty}</span>
      <button className="cart-item-update-qty-button" onClick={onIncrement}>
        +
      </button>
      <button className="cart-item-remove-button" onClick={onRemove}>
        Remove Item
      </button>
    </div>
  </div>
)

export default CartItem
