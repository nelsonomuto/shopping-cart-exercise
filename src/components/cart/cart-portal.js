import React from "react"
import { createPortal } from "react-dom"

const cartPortalRoot = document.body

class CartPortal extends React.Component {
  constructor(props) {
    super(props)
    this.el = document.createElement("div")
  }

  componentDidMount() {
    cartPortalRoot.appendChild(this.el)
  }

  componentWillUnmount() {
    cartPortalRoot.removeChild(this.el)
  }

  render() {
    return createPortal(this.props.children, this.el)
  }
}

export default CartPortal
