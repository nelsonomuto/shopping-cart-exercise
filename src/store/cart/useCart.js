import createContextStore from "../context-reducer-store"
import reducer, { getCartSummary } from "./reducer"

export const getDefaultState = (override = {}) => ({
  ...override,
  summary: getCartSummary(override.items || []),
  isCartOpen: false,
})

export const serializeStore = state => {
  return { items: state.items }
}

export const deserializeStore = fromStore => {
  return {
    items: fromStore.items,
    ...getDefaultState(fromStore),
  }
}

const {
  store,
  StoreProvider: CartStoreProvider,
  useStore: useCart,
} = createContextStore("Cart", reducer, getDefaultState(), {
  // only save `items` to localstorage, we dont need to save the whole cart state
  serializeStore,
  deserializeStore,
})

export { store }
export { CartStoreProvider }
export { useCart }
