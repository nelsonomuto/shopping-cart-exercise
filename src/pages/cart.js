import React from "react"
import "../assets/css/cart.css"
import CartList from "../components/cart/cart-list"
import CartSummary from "../components/cart/cart-summary"
import Layout from "../components/layout"
import SEO from "../components/seo"

const Cart = () => {
  return (
    <Layout>
      <SEO title="Cart" />
      <h1>Cart</h1>
      <CartList />
      <CartSummary />
    </Layout>
  )
}

export default Cart
