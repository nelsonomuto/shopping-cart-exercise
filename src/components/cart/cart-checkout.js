import React from "react"
import { dollars } from "../../services/formatString"
import { useCart } from "../../store/cart/useCart"
import { clearCart } from "../../store/cart/actions"
import ConfirmationButton from "../confirmation-button"
import CartList from "./cart-list"
import CartSummary from "./cart-summary"

const CartCheckout = () => {
  const [{ items, summary }, dispatch] = useCart()
  return (
    <>
      <CartList />
      <CartSummary />
      {items.length > 0 ? (
        <ConfirmationButton
          className="confirm-order-button"
          confirmText={`Confirm order for ${dollars(summary.totalCost)}`}
          cancelText="Cancel"
          onConfirm={() => {
            alert("Your order has been confirmed!")
            dispatch(clearCart())
          }}
        >
          Place Order
        </ConfirmationButton>
      ) : null}
    </>
  )
}

export default CartCheckout
