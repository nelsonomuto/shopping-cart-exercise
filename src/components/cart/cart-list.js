import { Link } from "gatsby"
import React from "react"
import { clearCart } from "../../store/cart/actions"
import { useCart } from "../../store/cart/useCart"
import CartItem from "./cart-item"

const CartList = ({ onShop }) => {
  const [{ items }, dispatch] = useCart()

  if (items.length === 0) {
    return (
      <div className="cart-list cart-list-empty">
        <h3>
          Your cart is empty.{" "}
          <Link to="/" onClick={onShop}>
            Shop now
          </Link>
        </h3>
      </div>
    )
  }

  return (
    <div className="cart-list-container">
      <div className="cart-list">
        {items.map(item => (
          <CartItem item={item} key={item.sku} />
        ))}
      </div>
      <button
        className="cart-list-clear-button"
        onClick={() => dispatch(clearCart())}
      >
        Clear Cart
      </button>
    </div>
  )
}

export default CartList
