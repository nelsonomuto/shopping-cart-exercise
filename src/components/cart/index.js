import { Link } from "gatsby"
import React, { useEffect, useRef } from "react"
import { dollars } from "../../services/formatString"
import { toggleCartOpen } from "../../store/cart/actions"
import { useCart } from "../../store/cart/useCart"
import CartList from "./cart-list"
import CartPortal from "./cart-portal"

const Cart = () => {
  const [{ isCartOpen, summary }, dispatch] = useCart()
  const containerRef = useRef(null)

  const toggleCart = () => dispatch(toggleCartOpen())

  const onClickOutside = e => {
    if (containerRef.current.contains(e.target)) {
      // inside
      return
    }
    // outside
    toggleCart()
  }

  useEffect(() => {
    if (isCartOpen) {
      document.addEventListener("mousedown", onClickOutside)
      return () => {
        document.removeEventListener("mousedown", onClickOutside)
      }
    } else {
      document.removeEventListener("mousedown", onClickOutside)
    }
  }, [isCartOpen])

  return (
    <CartPortal>
      <div
        onClick={onClickOutside}
        className={
          "cart-portal-container" +
          (isCartOpen ? " cart-animate-open" : " cart-animate-closed")
        }
      >
        <div className="cart-sidebar-container" ref={containerRef}>
          <button className="cart-portal-close-button" onClick={toggleCart}>
            Close
          </button>
          <div className="cart-sidebar-items-container">
            <h1>Cart</h1>
            <CartList onShop={toggleCart} />
          </div>
          <div className="cart-sidebar-summary-container">
            <div className="cart-subtotal">
              Subtotal: {dollars(summary.subtotal)}
            </div>
            <Link to="/checkout" className="cart-checkout-button">
              Checkout
            </Link>
          </div>
        </div>
      </div>
    </CartPortal>
  )
}

export default Cart
