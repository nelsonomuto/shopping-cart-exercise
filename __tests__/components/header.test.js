import { shallow } from "enzyme"
import React from "react"
import Header from "../../src/components/header"
import { useCart } from "../../src/store/cart/useCart"

const mockState = {
  items: [
    {
      sku: "fake-sku",
      title: "fake-title",
      price: 999.99,
      image: "fake-image-url",
      qty: 4,
    },
  ],
}

jest.mock("../../src/store/cart/useCart.js", () => ({
  useCart: jest.fn(),
}))

describe("<Header />", () => {
  it("matches snapshot", () => {
    useCart.mockImplementationOnce(() => [mockState])
    const component = shallow(<Header siteTitle="TEST" />)
    expect(component).toMatchSnapshot()
  })
})
