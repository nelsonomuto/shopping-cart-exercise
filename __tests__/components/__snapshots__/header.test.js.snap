// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Header /> matches snapshot 1`] = `
<header
  style={
    Object {
      "background": "rebeccapurple",
      "marginBottom": "1.45rem",
    }
  }
>
  <div
    style={
      Object {
        "margin": "0 auto",
        "maxWidth": 960,
        "padding": "0 1.0875rem",
      }
    }
  >
    <h1
      style={
        Object {
          "margin": 0,
          "position": "relative",
        }
      }
    >
      <mockConstructor
        style={
          Object {
            "color": "white",
            "textDecoration": "none",
          }
        }
        to="/"
      >
        TEST
      </mockConstructor>
      <button
        className="cart-button"
        onClick={[Function]}
      >
        CART (
        4
        )
      </button>
    </h1>
  </div>
</header>
`;
